### DISCUSSION TITLE
Userscripts not getting python modules from env

### DISCUSSION THREAD

**Discussion by CrowderSoup** [qutebrowser/qutebrowser#6134] [Q&A] [answered]
**Version info**:
qutebrowser v2.0.1
Git commit: 8fbd3c21a on HEAD (2021-01-28 17:43:33 +0100)
Backend: QtWebEngine (Chromium 83.0.4103.122)
Qt: 5.15.2

CPython: 3.9.1
PyQt: 5.15.2

sip: 5.4.0
colorama: 0.4.4
jinja2: 2.11.2
pygments: 2.7.4
yaml: 5.4.1
adblock: 0.4.1
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.2
PyQt5.QtWebKitWidgets: no
pdf.js: 2.6.347 (bundled)
sqlite: 3.33.0
QtNetwork SSL: Secure Transport, macOS 10.16

Style: QMacStyle
Platform plugin: cocoa
OpenGL: None
Platform: macOS-10.16-x86_64-i386-64bit, 64bit
Frozen: True
Imported from /Applications/qutebrowser.app/Contents/MacOS/qutebrowser
Using Python from /Applications/qutebrowser.app/Contents/MacOS/qutebrowser
Qt library executable path: /Applications/qutebrowser.app/Contents/MacOS/PyQt5/Qt/libexec, data path: /Applications/qutebrowser.app/Contents/MacOS/PyQt5/Qt
OS Version: 10.16, x86_64

**Does the bug happen if you start with `--temp-basedir`?**: Yes

**Description**

**How to reproduce**
When I attempt to spawn a userscript, I get errors saying "No Modules named <module>". I have tried several user scripts and it happens with all of them, and with a number of modules. I have the modules (`pip install PyNaCl` for instance), but I still get the error. This leads me to believe that the python environment my Qutebrowser is running in isn't the same as the environment that I'm installing the modules in.

![image](https://user-images.githubusercontent.com/521284/106966513-bfc6bb80-6702-11eb-91c9-d8f502b8f2ed.png)

Honestly I could just be doing this wrong but based on everything I've read this should be working.

**Comment by toofar** [comment]
> This leads me to believe that the python environment my Qutebrowser is running in isn't the same as the environment that I'm installing the modules in.

You didn't say how you are installing the modules but that may indeed be the case. The userscripts can be written in any language and are just executed like any old executable, so they don't inherit the interpreter or anything. I think there have been other reports about things like this being an extra pain on Macs, not sure if it is quite that same as what you are running into though (https://github.com/qutebrowser/qutebrowser/issues/4273)

**Comment by The-Compiler** [comment]
I'm converting this to a discussion - there's nothing special qutebrowser does here, like @toofar mentioned, those just run like normal executables using whatever Python interpreter is specified in their shebang.

**Comment by CrowderSoup** [comment, marked as answer]
I reviewed the ticket and related information linked by @toofar and I came up with a pretty decent solution so I figured I'd share it here.

The top of my `config.py` looks like this now:
```python
"""Qutebrowser configuration."""
import platform
import os
from os.path import expanduser

system = platform.system()
home_dir = expanduser('~')

if system == 'Darwin':
    # Add things to PATH so userscripts work
    path = os.environ['PATH']
    path += os.pathsep + home_dir + '/.pyenv/shims'
    path += os.pathsep + home_dir + '/.pyenv/plugins/pyenv-virtualenv/shims'
```

So now I can add things to my path as needed, and just need a quick `:config-source` to reload it. I am checking to see what `platform.system()` is because this is only an issue for me on OSX and I share this config between my personal Linux machine and my work Macbook.

**Reply by aleksejkrueger** [reply]
Hey,

could you please explain how you add things to the PATH?

When I try to append a path nothing happens.

```
"""Qutebrowser configuration."""
import platform
import os
from os.path import expanduser

system = platform.system()
home_dir = expanduser('~')

if system == 'Darwin':
    # Add things to PATH so userscripts work
    path = os.environ['PATH']
    path += os.pathsep + home_dir + '/.pyenv/shims'
    path += os.pathsep + home_dir + '/.pyenv/plugins/pyenv-virtualenv/shims'
    path += os.pathsep + '/usr/local/bin'
```

**Reply by mschilli87** [reply]
@alexchaichan: Are you on OSX? Else, the `system == Darwin` conditional will be `False` and the entire `path += ...` lines won't run.

**Reply by aleksejkrueger** [reply]
Yes I'm on osx.
On Linux I don't face this problems.

**Reply by toofar** [reply]
I'm pretty sure you need to assign it back to `os.environ['PATH']`. For example:
```
path = os.environ['PATH']
path += os.pathsep + '/usr/local/bin'
os.environ['PATH'] = path
```

Or just skip the `path` assignment and work on `os.environ['PATH']` directly, whatever is more clear to you.

**Reply by mschilli87** [reply]
@toofar: That would have been my next guess, too. :smile:

**Reply by aleksejkrueger** [reply]
Ahh thank you again!

assigning it back is the key.
